src/bootstrap/lib.rs

```diff
@@ -65,9 +65,15 @@
 //! also check out the `src/bootstrap/README.md` file for more information.
 
 #![deny(warnings)]
+#![feature(associated_consts)]
+#![feature(core_intrinsics)]
```

Comment:
A hashmap per T works if you generate methods for each T; that is, instead of the current approach of `builder.ensure` you'd have more of a `compiler::Rustc { ... }.ensure()` which would store itself onto the right hashmap in cache.